xcdat/regridder/regrid2.py

```diff
@@ -75,13 +89,21 @@ def horizontal(self, data_var: str, ds: xr.Dataset) -> xr.Dataset:
         dst_lat_bnds = _get_bounds_ensure_dtype(self._output_grid, "Y")
         dst_lon_bnds = _get_bounds_ensure_dtype(self._output_grid, "X")
 
-        src_mask_da = self._input_grid.get("mask", None)
+        if self._create_nan_mask:
+            non_yx = set(input_data_var.cf.axes.keys()) - set(["Y", "X"])
+            non_yx_selector: Dict[Any, Any] = {
+                input_data_var.cf[x].name: 0 for x in non_yx
+            }
 
-        # DataArray to np.ndarray, handle error when None
-        try:
-            src_mask = src_mask_da.values  # type: ignore
-        except AttributeError:
-            src_mask = None
+            src_mask = xr.where(
+                np.isnan(input_data_var.isel(**non_yx_selector)), 0, 1
+            ).values
+        else:
+            # DataArray to np.ndarray, handle error when None
+            try:
+                src_mask = self._input_grid.get("mask", None).values  # type: ignore
+            except AttributeError:
+                src_mask = create_mask(self._input_grid, ["Y", "X"]).values
```

Comment:
I think the `create_nan_mask` parameter and this logic should also be in the xESMF Regridder for more accurate regridding.